docs/BuildingLBANN.md

```diff
@@ -0,0 +1,382 @@
+
+The following LLNL-maintained packages are required.
+
++ [Hydrogen](https://github.com/llnl/elemental) is a fork of the
+  Elemental distributed dense linear-algebra library and it may be
+  installed via [spack](https://github.com/llnl/spack) using the
+  package name "hydrogen". If CUDA support is required in LBANN,
+  Hydrogen must also be build with CUDA support.
+
+The following LLNL-maintained packages are optional.
+
++ [Aluminum](https://github.com/llnl/aluminum) is a
+  communication library optimized for machine learning and interaction
+  with GPUs. We cannot recommend its use strongly enough. It can be
+  built using [spack](https://github.com/llnl/spack).
++ [CONDUIT](https://github.com/llnl/conduit) is used to ingest
+  structured data produced by scientific simulations.
+
+
+## Building with [Spack](https://github.com/llnl/spack)
+
+Some variation on the theme of `spack install lbann`. bvanessen should
+document the spack flags or whatever they call them that LBANN
+supports. I don't know the current state of spack with respect to its
+building LBANN correctly and/or successfully.
+
+## Buidling with [CMake](https://cmake.org)
+
+LBANN uses [CMake](https://cmake.org) for its build system and a
+version newer than or equal to 3.9.0 is required. LBANN development is
+done primarily on UNIX-based platforms. As such, the build is tested
+regularly on Linux-based machines, occasionally on OSX, and never on
+Windows machines.
+
+It is required that LBANN be built out-of-source. That is, CMake must
+not be invoked in a directory containing a CMakeLists.
+
+### LBANN CMake options
+The following options are exposed in the CMake build system.
+
++ `LBANN_WITH_ALUMINUM` (Default: `OFF`): Use the Aluminum communication
+  package. This will be set to `ON` automatically if Hydrogen was
+  built with Aluminum.
+
++ `LBANN_WITH_CNPY` (Default: `ON`): Build with support for CNPY for reading
+  Numpy data.
+
++ `LBANN_WITH_CUDA` (Default: `OFF`): Enable a CUDA-aware build. If
+  this flag is `ON` and Hydrogen CUDA support cannot be detected,
+  LBANN's configuration step will fail.
+
++ `LBANN_WITH_CONDUIT` (Default: `OFF`): Build with support for CONDUIT.
+
++ `LBANN_WITH_NVPROF` (Default: `OFF`): Build with extra annotations for NVPROF.
+
++ `LBANN_WITH_TOPO_AWARE` (Default: `ON`): Use HWLOC for topology-aware choices.
+
++ `LBANN_WITH_TBINF` (Default: `ON`): Enable the Tensorboard interace.
+
++ `LBANN_WITH_VTUNE` (Default: `OFF`): Build with extra annotations for VTune.
+
++ `LBANN_DETERMINISTIC` (Default: `OFF`): Force as much of the code as possible
+  to be deterministic. This is not a guarantee as certain operations
+  in third-party libraries cannot be forced into a deterministic mode,
+  especially for CUDA-enabled builds.
+
++ `LBANN_SEQUENTIAL_INITIALIZATION` (Default: `OFF`): Force sequentially
+  consistent initialization of data structures.
+
++ `LBANN_WARNINGS_AS_ERRORS` (Default: `OFF`): Promote compiler
```

Comment:
If all developers build with this on, there should be no issues with users building with this on. Therefore it should always be on. 😉

Reply:
My concern is overly enthusiastic users (ok, probably just me) using weird compilers at newer or older versions than we test on. E.g., clang 6+ complains about the way El/Hydrogen abuse templates in the MPI wrapper and GCC does not. Then there's all the weird stuff Intel emits. I don't want to be on the hook for appeasing all user-end compilers that dislike something in our style. But if you're volunteering... :)